src/compiler/checker.ts

```diff
@@ -14169,9 +14175,11 @@ namespace ts {
                     (type === falseType || type === regularFalseType) ? TypeFacts.FalseFacts : TypeFacts.TrueFacts;
             }
             if (flags & TypeFlags.Object) {
-                return isFunctionObjectType(<ObjectType>type) ?

```

Comment:
The empty object type is a red herring. The issue also exists for other types which have falsey types structurally assignable to them:, like `{ toString(): string; }` or `{ toFixed(): string; }`.

Reply:
Right, in an ideal world we'd check whether `""`, `0`, or `false` is assignable to the type in question, but I'm concerned about the performance impact.

Reply:
I think it shouldn't be _too_ bad, given that we'll be doing the check only once for all string-subtypes, number-subtypes, etc vs each input type, then caching the relation comparison result for every other time we check it. And the check itself should (in the common case of failure) bail out really early when the property keys don't sufficiently overlap.